anx/provider/loadbalancer/reconciliation/state_retriever.go

```diff
@@ -0,0 +1,411 @@
+		tags:          []string{serviceTag},
+	}
+
+	// initialize stateRetrieverLoadBalancerContainer for each loadbalancer
+	for _, lbIdentifier := range lbIdentifiers {
+		sr.loadBalancers[lbIdentifier] = &stateRetrieverLoadBalancerContainer{
+			updateComplete: make(chan interface{}),
+			state:          &remoteLoadBalancerState{},
+		}
+	}
+
+	// add all registered loadbalancers to waitgroup (used to sync state retrieving)
+	sr.loadBalancersReadyWaitGroup.Add(len(lbIdentifiers))
+
+	// start internal update loop
+	go sr.updateLoop(ctx)
+
+	return sr
+}
+
+// Done unregisters a loadbalancer from the state retriever.
+// The stateRetriever won't wait for the LoadBalancerState(lbIdentifier) call in future state retrieving iterations.
+func (r *stateRetrieverImpl) Done(lbIdentifier string) error {
+	r.loadBalancersMapLock.Lock()
+	defer r.loadBalancersMapLock.Unlock()
+
+	if _, ok := r.loadBalancers[lbIdentifier]; !ok {
+		return errLoadBalancerNotRegistered
+	}
+
+	close(r.loadBalancers[lbIdentifier].updateComplete)
+	delete(r.loadBalancers, lbIdentifier)
+	r.loadBalancersReadyWaitGroup.Done()
+
+	return nil
+}
+
+// LoadBalancerState returns the loadbalancer state filtered by loadbalancer ID
+func (r *stateRetrieverImpl) LoadBalancerState(lbIdentifier string) (*remoteLoadBalancerState, error) {
+	loadBalancer, err := r.getLoadBalancer(lbIdentifier)
+	if err != nil {
+		return nil, err
+	}
+
+	// lock LoadBalancerState for lbIdentifier
+	loadBalancer.lock.Lock()
+	defer loadBalancer.lock.Unlock()
+
+	r.loadBalancersReadyWaitGroup.Done()
+
+	// wait until all registered lb's called LoadBalancerState
+	<-loadBalancer.updateComplete
+
+	if r.err != nil {
+		return nil, r.err
+	}
+
+	return loadBalancer.state, nil
+}
+
+// updateLoop is an internal update loop called exactly once for each state retriever instance.
+// It is responsible for waiting for all loadbalancers to be ready for the next state retrival,
+// updating the state and finally start over until all loadbalancers are Done().
+func (r *stateRetrieverImpl) updateLoop(ctx context.Context) {
+	waitChannel := make(chan interface{})
+	defer close(waitChannel)
+
+	for {
+		go func() { // make waitgroup "selectable" (allows updateLoop to be cancelable via ctx)
+			r.loadBalancersReadyWaitGroup.Wait()
+			waitChannel <- nil
+		}()
+
+		select {
+		case <-waitChannel: // all loadbalancers ready or done
+			break
+		case <-ctx.Done(): // context expired
+			r.err = ctx.Err()
+			logr.FromContextOrDiscard(ctx).Error(ctx.Err(), fmt.Sprintf("LoadBalancer state retriever for %q didn't finish properly", r.tags[0]))
+			for _, lb := range r.loadBalancers {
+				close(lb.updateComplete)
+			}
+			return
+		}
+
+		r.loadBalancersMapLock.Lock()
+		if len(r.loadBalancers) == 0 {
+			// all done
+			return
+		}
+
+		// retrieve remote state
+		r.update(ctx)
+
+		r.loadBalancersReadyWaitGroup.Add(len(r.loadBalancers))
+
+		for _, lb := range r.loadBalancers {
+			lb.updateComplete <- nil
+		}
```

Comment:
please check if we can use a `sync.Cond` instead and `.Broadcast()` it here or at least only have a single channel for all to wait on (closing the channel is the `Broadcast` with channels)